### PR TITLE
Implements a two-phase parse for expressions

### PR DESCRIPTION
**Pull request by psiservices-dbrink**
ANTLR4's default prediction mode is more powerful, but slower.  This PR implements a two-phase parse, the first using the faster SLL mode, falling back to the slower, but more powerful LL mode on failure.

This may address #78 

See https://medium.com/pragmatic-programmers/maximizing-parser-speed-9766aaaab0c0

### CODE REVIEW

**Comment by psiservices-dbrink** [comment]
Results appear mixed.

Before change:
https://ci.appveyor.com/project/sklose/ncalc2/builds/48865720

> ------------------------------------------------------------
> Formula: 5 * 2 = 2 * 5 && (1 / 3.0) * 3 = 1
> Expression: 158,863.48 evaluations / sec
> Lambda: 60,945,880.06 evaluations / sec
> Lambda Speedup: 38,263.68%%
> ------------------------------------------------------------
> ------------------------------------------------------------
> Formula: (4 * 12 / 7) + ((9 * 2) % 8)
> Expression: 234,288.83 evaluations / sec
> Lambda: [55](https://ci.appveyor.com/project/sklose/ncalc2/builds/48865720#L55),900,273.91 evaluations / sec
> Lambda Speedup: 23,759.[56](https://ci.appveyor.com/project/sklose/ncalc2/builds/48865720#L56)%%
> ------------------------------------------------------------
> ------------------------------------------------------------
> Formula: Foo([Param1] * 7, [Param2])
> Expression: 351,821.43 evaluations / sec
> Lambda: 99,088,386.84 evaluations / sec
> Lambda Speedup: 28,064.40%%
> ------------------------------------------------------------
> ------------------------------------------------------------
> Formula: [Param1] * 7 + [Param2]
> Expression: 407,567.39 evaluations / sec
> Lambda: 101,646,676.15 evaluations / sec
> Lambda Speedup: 24,839.85%%
> ------------------------------------------------------------
> ------------------------------------------------------------
> Formula: [Param1] * 7 + [Param2]
> Expression: 4[61](https://ci.appveyor.com/project/sklose/ncalc2/builds/48865720#L61),207.39 evaluations / sec
> Lambda: 93,449,210.35 evaluations / sec
> Lambda Speedup: 20,161.86%%
> ------------------------------------------------------------

After change:
https://ci.appveyor.com/project/sklose/ncalc2/builds/48866282

> 
> ------------------------------------------------------------
> Formula: 5 * 2 = 2 * 5 && (1 / 3.0) * 3 = 1
> Expression: 138,617.39 evaluations / sec
> Lambda: 66,229,[55](https://ci.appveyor.com/project/sklose/ncalc2/builds/48866282#L55)1.63 evaluations / sec
> Lambda Speedup: 47,678.67%%
> ------------------------------------------------------------
> ------------------------------------------------------------
> Formula: (4 * 12 / 7) + ((9 * 2) % 8)
> Expression: 262,643.11 evaluations / sec
> Lambda: 74,161,969.74 evaluations / sec
> Lambda Speedup: 28,136.78%%
> ------------------------------------------------------------
> ------------------------------------------------------------
> Formula: Foo([Param1] * 7, [Param2])
> Expression: 491,892.14 evaluations / sec
> Lambda: 102,923,013.59 evaluations / sec
> Lambda Speedup: 20,823.90%%
> ------------------------------------------------------------
> ------------------------------------------------------------
> Formula: [Param1] * 7 + [Param2]
> Expression: 706,603.92 evaluations / sec
> Lambda: 117,980,179.33 evaluations / sec
> Lambda Speedup: 16,596.79%%
> ------------------------------------------------------------
> ------------------------------------------------------------
> Formula: [Param1] * 7 + [Param2]
> Expression: 819,865.00 evaluations / sec
> Lambda: 106,791,969.24 evaluations / sec
> Lambda Speedup: 12,925.[56](https://ci.appveyor.com/project/sklose/ncalc2/builds/48866282#L56)%%
> ------------------------------------------------------------

**Comment by Bykiev** [comment]
Can you please split the changes into 2 separate PRs? Also the benchmark with the changes will be highly appreciated

File: src/NCalc/Expression.cs [hunk 1]
```diff
@@ -121,11 +121,28 @@ public static LogicalExpression Compile(string expression, bool nocache)
...
+
+                try
+                {
+                    logicalExpression = parser.ncalcExpression().value;
+                }
+                catch(ParseCanceledException)
+                {
+                    lexer.Reset();
+
+                    parser.Reset();
```
Comment (sklose) [inline src/NCalc/Expression.cs:140]: I am curious if/when this fallback mechanism is hit. Have you seen this being used for any of the test cases?

Comment (psiservices-dbrink) [reply src/NCalc/Expression.cs:140]: The code is hit any time there is a parser error.  The test ShouldDetectSyntaxErrorsBeforeEvaluation hits the catch, but it doesn't test a successful fallback to the LL prediction mode.   I tried several complex expressions, but I wasn't able to find anything that failed on SLL and then succeeded on LL.  I'm not convinced the NCalc grammar is complex enough to need anything more than SLL, but I don't know how to prove it.  

The SLL parser is generally faster, and the LL fallback seems prudent.

**Pull request merged**